<h5 style="color:#2dede4;">
  📑 Projeto – Controle de Contas a Pagar e a Receber
</h5>

<p style="font-size: 14px">
Neste exercício, você irá desenvolver um programa em <strong>Python</strong> que simula
um <strong>sistema simples de controle financeiro corporativo</strong>, responsável por
processar contas a pagar e a receber a partir de um arquivo de dados.
</p>

<p style="font-size: 14px">
O sistema irá realizar a leitura de um arquivo <strong>JSON</strong>, aplicar regras básicas
de vencimento e gerar análises financeiras que auxiliam na tomada de decisão.
</p>

<hr>

<p style="font-size: 18px; color:#2dede4">
  🎯 Objetivo
</p>
<p style="font-size: 14px">Criar um programa que:</p>
<ul style="font-size: 14px">
  <li>Leia contas financeiras a partir de um arquivo JSON</li>
  <li>Identifique contas a pagar e a receber</li>
  <li>Verifique contas vencidas com base na data atual</li>
  <li>Calcule totais financeiros (pagar, receber e vencidos)</li>
  <li>Exiba um resumo financeiro no terminal</li>
</ul>

<hr>

<p style="font-size: 18px; color:#2dede4">
  📥 Dados de Entrada
</p>
<p style="font-size: 14px">
O programa deverá consumir um arquivo <code>contas.json</code> contendo informações como:
</p>
<ul style="font-size: 14px">
  <li>ID da conta</li>
  <li>Tipo (Pagar ou Receber)</li>
  <li>Descrição</li>
  <li>Valor</li>
  <li>Data de vencimento</li>
  <li>Status da conta</li>
</ul>

<hr>

<p style="font-size: 18px; color:#2dede4">
  🕒 Regras de Funcionamento
</p>

<table border="1" cellpadding="8" cellspacing="0">
  <thead style="background-color:#e0f2f1;">
    <tr style="font-size: 14px; color:#000">
      <th>Regra</th>
      <th>Descrição</th>
    </tr>
  </thead>
  <tbody style="font-size: 14px">
    <tr>
      <td>Conversão de data</td>
      <td>A data de vencimento deve ser convertida para o tipo <code>datetime</code></td>
    </tr>
    <tr>
      <td>Conta vencida</td>
      <td>Conta em aberto com vencimento anterior à data atual</td>
    </tr>
    <tr>
      <td>Saldo a pagar</td>
      <td>Soma dos valores das contas do tipo <strong>Pagar</strong></td>
    </tr>
    <tr>
      <td>Saldo a receber</td>
      <td>Soma dos valores das contas do tipo <strong>Receber</strong></td>
    </tr>
  </tbody>
</table>

<br>

<p style="font-size: 18px; color:#2dede4">
  📊 Processamento dos Dados
</p>

<table border="1" cellpadding="8" cellspacing="0">
  <thead style="background-color:#e0f2f1;">
    <tr style="font-size: 14px; color:#000">
      <th>Etapa</th>
      <th>Descrição</th>
    </tr>
  </thead>
  <tbody style="font-size: 14px">
    <tr>
      <td>Leitura</td>
      <td>Carregar os dados do arquivo JSON</td>
    </tr>
    <tr>
      <td>Tratamento</td>
      <td>Converter tipos de dados e organizar as informações</td>
    </tr>
    <tr>
      <td>Análise</td>
      <td>Identificar contas vencidas e calcular totais financeiros</td>
    </tr>
    <tr>
      <td>Resumo</td>
      <td>Exibir os resultados consolidados no terminal</td>
    </tr>
  </tbody>
</table>

<hr>

<p style="font-size: 18px; color:#2dede4">
  📤 Saída Esperada
</p>
<p style="font-size: 14px">Ao final da execução, o programa deve exibir:</p>
<ul style="font-size: 14px">
  <li>Total de contas a pagar</li>
  <li>Total de contas a receber</li>
  <li>Total de contas vencidas</li>
  <li>Saldo financeiro consolidado</li>
</ul>

<p style="background:#f1f8e9; padding:2px; border-left:6px solid #000000; color:#000051; font-size:14px">
💡 <strong>Dica:</strong> utilize <strong>funções</strong> para organizar o código,
o módulo <strong>datetime</strong> para trabalhar com datas
e estruturas como <strong>listas</strong> e <strong>dicionários</strong> para armazenar os dados.
</p>

<p style="text-align:center; color:#00ff37;">
🚀 Bora programar! 👽
</p>



Importando bibliotecas

In [43]:
from datetime import datetime
import json

Carregando arquivo

In [44]:
caminho = 'dados\contas.json'

try:
    with open(caminho, 'r', encoding="UTF8") as arquivo:
        contas = json.load(arquivo)

except FileNotFoundError as e:
    print(e)

Tratamento de Dados - Convertendo Datas

In [45]:
# Convertendo a data texto em data real
for conta in contas:
    conta['vencimento'] = datetime.strptime(conta['vencimento'], "%d/%m/%Y")

# Verificando a conversão
contas[0]['vencimento']

datetime.datetime(2026, 1, 27, 0, 0)

Contas Vencidas

In [46]:
contas_vencidas = []
contas_vencidas_pagar = []
contas_vencidas_receber = []
contas_erro_tipo = []

for conta in contas:
    if conta['vencimento'] < datetime.today():
        contas_vencidas.append(conta)

        if conta['tipo'] == 'Pagar':
            contas_vencidas_pagar.append(conta)

        elif conta['tipo'] == 'Receber':
            contas_vencidas_receber.append(conta)
        
        else:
            contas_erro_tipo.append(conta)

print(len(contas_vencidas))
print(len(contas_vencidas_pagar))
print(len(contas_vencidas_receber))
print(len(contas_erro_tipo))

3000
1549
1450
1


In [38]:
saldo_vencido = 0

for conta in contas:
    if conta['vencimento'] < datetime.today() and conta['tipo'] == 'Pagar'and conta['status'] == 'Em aberto':
        saldo_vencido += conta['valor']

print(f"Saldo de Contas à Pagar Vencidas: R$ {saldo_vencido: ,.2f}")

Saldo de Contas à Pagar Vencidas: R$  1,965,984.26


Contas à Pagar

In [24]:
saldo_a_pagar = 0

for conta in contas:
    if conta['tipo'] == 'Pagar':
        saldo_a_pagar += conta['valor']

print(f"Total de Contas à Pagar: R$ {saldo_a_pagar: ,.2f}")

Total de Contas à Pagar: R$  3,997,497.53


Contas à Receber

In [29]:
saldo_a_receber = 0

for conta in contas:
    if conta['tipo'] == 'Receber':
        saldo_a_receber += conta['valor']

print(f"Total de Contas à Receber: R$ {saldo_a_receber: ,.2f}")

Total de Contas à Receber: R$  3,635,999.92


Resultado Consolidado

In [35]:
#contas_pagas = saldo_a_pagar - saldo_vencido

#saldo_final1 = saldo_a_receber - contas_pagas 
#saldo_final2 = saldo_a_receber - contas_pagas - saldo_vencido

saldo_final = saldo_a_receber - saldo_a_pagar

#print(f'Saldo antes do pagamento das contas vencidas: R$ {saldo_final1: ,.2f}')
#print(f'O saldo final após pagamento de contas vencidas: R$ {saldo_final2: ,.2f}')
print(f'O saldo final é: R$ {saldo_final: ,.2f}')

O saldo final é: R$ -361,497.61


# Resumo

De acordo com as análises realizadas, chegamos as seguintes conclusões:

<ul style="font-size: 14px">
  <li>Total de contas a pagar: R$ 3,997,497.53</li>
  <li>Total de contas a receber: R$ 3,635,999.92 </li>
  <li>Total de contas vencidas: R$ 1,965,984.26</li>
  <li>Saldo financeiro consolidado: R$ -361,497.61 </li>
</ul>
